consensus/polybft/proposer_calculator.go

```diff
@@ -0,0 +1,499 @@
+	// current snapshot
+	snapshot *ProposerSnapshot
+
+	// logger instance
+	logger hclog.Logger
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculator(config *runtimeConfig, logger hclog.Logger) (*proposerCalculator, error) {
+	snap, err := NewProposerSnapshotFromState(config, logger)
+
+	if err != nil {
+		return nil, err
+	}
+
+	return &proposerCalculator{
+		snapshot: snap,
+		logger:   logger,
+	}, nil
+}
+
+// NewProposerCalculator creates a new proposer calculator object
+func NewProposerCalculatorFromSnapshot(pcs *ProposerSnapshot, logger hclog.Logger) *proposerCalculator {
+	return &proposerCalculator{
+		snapshot: pcs.Copy(),
+		logger:   logger,
+	}
+}
+
+// Get copy of the proposers' snapshot
+func (pc *proposerCalculator) GetSnapshot() (*ProposerSnapshot, bool) {
+	if pc.snapshot == nil {
+		return nil, false
+	}
+
+	return pc.snapshot.Copy(), true
+}
+
+// GetLatestProposer returns latest calculated proposer if any
+func (pc *proposerCalculator) GetLatestProposer(round, height uint64) (types.Address, bool) {
+	// round must be same as saved one and proposer must exist
+	if pc.snapshot == nil || pc.snapshot.Proposer == nil || pc.snapshot.Round != round || pc.snapshot.Height != height {
+		pc.logger.Info("Get latest proposer not found", "height", height, "round", round,
+			"pc height", pc.snapshot.Height, "pc round", pc.snapshot.Round)
+
+		return types.ZeroAddress, false
+	}
+
+	pc.logger.Info("Get latest proposer",
+		"height", height, "round", round, "address", pc.snapshot.Proposer.Metadata.Address)
+
+	return pc.snapshot.Proposer.Metadata.Address, true
+}
+
+// CalcProposer calculates next proposer
+func (pc *proposerCalculator) CalcProposer(round, height uint64) (types.Address, error) {
+	if height != pc.snapshot.Height {
+		return types.ZeroAddress, fmt.Errorf("invalid height - expected %d, got %d", pc.snapshot.Height, height)
+	}
+
+	// optimization -> return current proposer if already calculated for this round
+	if pc.snapshot.Round == round && pc.snapshot.Proposer != nil {
+		return pc.snapshot.Proposer.Metadata.Address, nil
+	}
+
+	// do not change priorities on original snapshot while executing CalcProposer
+	// if round = 0 then we need one iteration
+	proposer, err := incrementProposerPriorityNTimes(pc.snapshot.Copy(), round+1)
+	if err != nil {
+		return types.ZeroAddress, err
+	}
+
+	pc.snapshot.Proposer = proposer
+	pc.snapshot.Round = round
+
+	pc.logger.Info("New proposer calculated", "height", height, "round", round, "address", proposer.Metadata.Address)
+
+	return proposer.Metadata.Address, nil
+}
+
+func (pc *proposerCalculator) Update(blockNumber uint64, config *runtimeConfig, state *State) error {
+	const saveEveryNIterations = 5
+
+	pc.logger.Info("Update proposal snapshot started", "block", blockNumber)
+
+	from := pc.snapshot.Height
+
+	for height := from; height <= blockNumber; height++ {
+		if err := pc.updatePerBlock(height, config); err != nil {
```

Comment:
This is just a question but could not we ensure that this gets called on a sequence of blocks and we do not have sequences of skipped blocks?

Reply:
This is actually called whenever a block is inserted (`OnBlockInserted`). That `FOR` loop should do only one iteration each time because the snapshot of the previous block was (hopefully) saved.  
We made it with the `FOR` loop just in case that somehow the snapshot of the previous block was not saved (really unexpected situation occurred). That way snapshot would catch up the blockchain.